### PR TITLE
Update ScalarSymbolicOps and trainability

### LINKED ISSUE #3916: [BUG] Composite operator parameters are not trainable
### Expected behavior

The parameters of composite ops are trainable (just like the 1st-generation Tensor and Hamiltonians).

### Actual behavior

They are not trainable.

### Additional information

This is because composite operators put their parameters in vanilla python lists, and those are considered not trainable. My code snippet first shows a full example (mimicking [how pennylane currently sets trainable parameters](https://github.com/PennyLaneAI/pennylane/blob/601c5ca6f5caec9c917c2c57fb5b61f6ed5f4ffa/pennylane/interfaces/autograd.py#L72-L73)), and then a minimal demonstration of what we currently consider trainable.

### Source code

```shell
dev = qml.device("default.qubit", wires=2)
@qml.qnode(dev)
def circuit():
    return qml.expval(qml.prod(qml.PauliX(0), qml.RX(qnp.array(1.1), 1)))

circuit()
params = circuit.tape.get_parameters(trainable_only=False)
print(params, qml.math.get_trainable_indices(params))
# [[], [tensor(1.1, requires_grad=True)]] set()


### Minimal demo of why it doesn't work
from pennylane.math import requires_grad; from pennylane import numpy as np
print(requires_grad(np.array(1)), requires_grad([np.array(1)]))
# True False
```

### Tracebacks

_No response_

### System information

```shell
Name: PennyLane
Version: 0.30.0.dev0
Summary: PennyLane is a Python quantum machine learning library by Xanadu Inc.
Home-page: https://github.com/XanaduAI/pennylane
Author: 
Author-email: 
License: Apache License 2.0
Location: /Users/matthews/.pyenv/versions/3.9.13/envs/pl/lib/python3.9/site-packages
Editable project location: /Users/matthews/src/github.com/PennyLaneAI/pennylane
Requires: appdirs, autograd, autoray, cachetools, networkx, numpy, pennylane-lightning, requests, retworkx, scipy, semantic-version, toml
Required-by: PennyLane-Lightning

Platform info:           macOS-13.2.1-arm64-arm-64bit
Python version:          3.9.13
Numpy version:           1.23.5
Scipy version:           1.10.0
Installed devices:
- default.gaussian (PennyLane-0.30.0.dev0)
- default.mixed (PennyLane-0.30.0.dev0)
- default.qubit (PennyLane-0.30.0.dev0)
- default.qubit.autograd (PennyLane-0.30.0.dev0)
- default.qubit.jax (PennyLane-0.30.0.dev0)
- default.qubit.tf (PennyLane-0.30.0.dev0)
- default.qubit.torch (PennyLane-0.30.0.dev0)
- default.qutrit (PennyLane-0.30.0.dev0)
- null.qubit (PennyLane-0.30.0.dev0)
- lightning.qubit (PennyLane-Lightning-0.29.0.dev11)
```

### Existing GitHub issues

- [X] I have searched existing GitHub issues to make sure the issue does not already exist.

### PR DESCRIPTION
**Pull request by timmysilv**
**Context:**
We want all operator data to be as flat as possible. Also, I had previously done this for Sum and Prod, but didn't test the consequential result which is that their data can now be considered trainable.

**Description of the Change:**
1. Move identical data logic from SProd and Exp to the parent class, ScalarSymbolicOp. Note that while doing this, I also made the data for `Exp` flat.
2. Add a custom `data` property definition to Pow so it keeps its old behaviour of treating `z` as a hyperparameter (eg. excluded from `op.data`)
3. The `copy` dunder for Exp was moved to the parent class so its siblings can have it too : )
4. Add tests to ensure Exp, Sum and Prod data is trainable (there was already an SProd test that actually measures the gradient)

**Benefits:**
- It is easier to detect trainability of flat data, and more consistent with the rest of PennyLane
- Less copied code
- The new tape converter integration can move forward thanks to the addition of a data setter to all ScalarSymbolicOps

**Possible Drawbacks:**
N/A

**Related GitHub Issues:**
#3958, #3899, fixes #3916

### CODE REVIEW

File: pennylane/ops/op_math/exp.py [hunk 1]
```diff
@@ -209,6 +209,10 @@ def hash(self):
     def data(self):
         return self._data
 
+    @data.setter
+    def data(self, new_data):
+        self._data = new_data
```
Comment (albi3ro) [inline pennylane/ops/op_math/exp.py:214]:
```suggestion
        self._data = new_data
        self.base.data = new_data[1:]
```

File: tests/ops/op_math/test_exp.py [hunk 2]
```diff
@@ -111,7 +111,13 @@ def test_data(self):
...
         op = Exp(base, coeff)
 
-        assert op.data == [[coeff], [phi]]
+        assert op.data == [coeff, phi]
+
+        new_phi = np.array(0.1234)
+        new_coeff = np.array(3.456)
+        op.data = [new_phi, new_coeff]
+
+        assert op.data == [new_phi, new_coeff]
```
Comment (timmysilv) [inline tests/ops/op_math/test_exp.py:120]:
```suggestion
        op.data = [new_coeff, new_phi]

        assert op.data == [new_coeff, new_phi]
        assert op.base.data == [new_phi]
        assert op.scalar == new_coeff
```

File: tests/ops/op_math/test_evolution.py [hunk 3]
```diff
@@ -83,6 +83,13 @@ def test_data(self):
         assert op.coeff == 1j * op.data[0]
         assert op.param == op.data[0]
 
+        new_param = np.array(2.345)
+        op.data = [new_param]
+
+        assert op.data == [new_param]
+        assert op.coeff == 1j * op.data[0]
+        assert op.data == op.data[0]
```
Comment (mudit2812) [inline tests/ops/op_math/test_evolution.py:91]: Are there any tests for using the setter for invalid data? What would happen if I set `op.data = np.array([1.2, 2.3])`? Would the operator then be considered broadcasted or would it break something?

File: tests/ops/op_math/test_exp.py [hunk 4]
```diff
@@ -44,8 +44,8 @@ def test_pauli_base(self, constructor):
         assert op.id == "something"
 
         assert op.num_params == 1
-        assert op.parameters == [[1], []]
-        assert op.data == [[1], []]
+        assert op.parameters == [1]
+        assert op.data == [1]
```
Comment (mudit2812) [inline tests/ops/op_math/test_exp.py:48]: This looks so much nicer than before. Kudos!

**Review by mudit2812 — COMMENTED** [review]
Thanks for the changes Matt! Glad to see trainability working with the flat data changes. Do tests for trainability of `Exp.data` need to be added too?

File: tests/ops/op_math/test_evolution.py (hunk 3 above)
Comment (timmysilv) [reply tests/ops/op_math/test_evolution.py:91]: I didn't add tests because we don't have a precedent for this. The rule is "don't do it".

Example:
```pycon
>>> op = qml.PauliX(0); op.data = 123; op.data
123
```

We can ensure the data is well-formed in all data setters if we feel it's important. Personally, I'd rather focus on making immutable operators.. but short-term this will be very helpful.

File: pennylane/ops/op_math/exp.py [hunk 5]
```diff
@@ -209,6 +209,12 @@ def hash(self):
...
+        self.scalar = new_data[0]
+        self.base.data = new_data[1:]
+
```
Comment (albi3ro) [inline pennylane/ops/op_math/exp.py:217]: Can we put this logic into `ScalarSymbolicOp` instead? Then we can remove it from `SProd` as well.

Also, the `_data` thing was to try and allow us to be able to dynamically update `self.base.data`. We may not need `_data` anymore.

Comment (timmysilv) [reply pennylane/ops/op_math/exp.py:217]: interesting.. on master we can't update `self.data`, and if we update `self.base.data` directly, it isn't propagated to `self.data` which isn't great. so in the current state of this PR, `SProd` and `Exp` are identical except that `Exp` caches the data for reads.

I'd like to note generally that property setters do not handle indexing (eg `op.data[1] = 1.1`) - it reads `op.data`, then modifies what it reads. With the new converter, this won't matter. But with today's `Unwrap`, this means ops like this are never set properly.

`Pow` also inherits from `ScalarSymbolicOp`, but its scalar is not part of its data. Should I change that in this PR as well? Otherwise I don't want to move it to the parent class.

Comment (timmysilv) [reply pennylane/ops/op_math/exp.py]: update: I did move it to the parent class, but overrode Pow's data behaviour since it is the outlier here. We can remove the Pow overrides if/when we want them to be the same.

**Review by albi3ro — APPROVED** [review]
Nice to have this all standardized :)

File: pennylane/ops/op_math/pow.py [hunk 6]
```diff
@@ -236,6 +236,15 @@ def z(self):
     def ndim_params(self):
         return self.base.ndim_params
 
+    @property
+    def data(self):
+        """The trainable parameters"""
+        return self.base.data
```
Comment (mudit2812) [inline pennylane/ops/op_math/pow.py:242]: This seems like an extremely stupid question to ask, but is `z` _never_ supposed to be trainable?

**Review by mudit2812 — APPROVED** [review]
Fairly certain that my question below is dumb, but regardless, this looks good to go

**Comment by mudit2812** [comment]
🚀

File: pennylane/ops/op_math/pow.py (hunk 6 above)
Comment (timmysilv) [reply pennylane/ops/op_math/pow.py:242]: I was wondering something similar in the above thread, but I'm not gonna question it in this PR.

Comment (albi3ro) [reply pennylane/ops/op_math/pow.py:242]: This allowed the base data to also be updated when `Pow`'s data was updated. We can reconsider this decision once parameter-shift no longer uses `QuantumScript.set_parameters` and in-place manipulation of an array.
